rfcs/0166-nix-formatting.md

```diff
@@ -0,0 +1,1059 @@
+**Rationale and alternatives**
+
+- Have leading commas for parameters in attribute set arguments, like currently done in Nixpkgs
+  ```nix
+  #6
+  { some
+  , arg
+  }:
+  #7
+  args@{
+    some
+  , argument
+  # Single line comment
+  , commentedArgument
+  , # Comment on the value
+    # multiline comment
+    default ? value
+  , ...
+  }:
+  # …
+  ```
+  - This leads to problems with the first argument, as leading commas are not allowed. `{ some` is discouraged by the the style guidelines; `some` should start on a new line instead. Also, this does not work well with `@` bindings.
+  - The currently suggested style for commenting items in the Nixpkgs manual (depicted here in `#7`) is not great. However, there are no other good solutions with leading comma style that don't run into other problems.
+  - The leading comma style was a lesser-evil workaround for the lack of trailing commas in the Nix language. Now that the language has this feature, there is no reason to keep it that way anymore.
+
+
+### Operators
+
+**Description**
+
+Chained binary associative [operators](https://nixos.org/manual/nix/stable/language/operators.html#operators) (except [function application](#function-application)) with the same or monotonically decreasing precedence are treated as one.
+
+If an operator chain does not fit onto one line, it is expanded such that every operator starts a new line:
+- If the operand can also fit on the same line as the operator, it's put there
+- Otherwise, the operand usually starts indented on a new line, with special handling for parenthesis, brackets, braces, function applications
+
+Operator chains may be compacted as long as all lines between the first and last one are indented.
+
+**Examples**
+
+```nix
+# These chained associative operators have increasing precedence, so they're _not_ treated the same
+foo
+-> # <- The operator starts on a new line, but right operand is all of the below lines, they don't fit here, so indent
+  bar
+  ||
+    baz
+    && qux # <- The operand fits on this line
+
+# These chained associative operators have decreasing precedence, so they're treated the same
+foo
+&& bar # <- All of these operands are just identifiers, they fit on the same line
+|| baz # <- We shouldn't indent these lines, because it misleads into thinking that || binds stronger than &&
+-> qux
+
+[
+  some
+  flags
+]
+++ ( # <- The operator is on a new line, but parenthesis/brackets/braces can start on the same line
+  foo
+)
+++ optionals condition [ # <- As are multiline functions applications
+  more
+  items
+]
+++
+  runCommand name # <- But only functions whose last argument can start on the same line
+    ''
+      echo hi
+    ''
+    test
+
+# We can use a more compact form as long as all inbetween lines are indented
+bar // {
+  x = 10;
+  y = 20;
+} // baz
+
+# Bad, we can't use the more compact form because an intermediate line is not indented.
+{
+  x = 10;
+  y = 20;
+} // foo // {
+  z = 30;
+  w = 40;
+}
+```
```

Comment:
That's not clear to me intuitively, and only makes sense if you follow the "intermediate lines must be indented" rule literally. A "good" alternative would help here. Is that okay, since sequenced operators start on the same level?

```nix
{
  x = 10;
  y = 20;
} // foo
// {
  z = 30;
  w = 40;
}
```

It doesn't feel right.

Reply:
Ah we just pushed something earlier that made this not quite work out. Notably this compact form should be specific to binders. I undid it again and added another example in https://github.com/NixOS/rfcs/pull/166/commits/940312fbdf41a89e4fdd3756f3724bc9cabd1d6f.

We think that making operators be on the same indentation level would be most consistent, but the compact form is there to make cases like these still work:

```nix
{
  foo = foo // {
    # Only 1 extra indentation level here
    bar = 10;
  };
}
```

The intermediate indentation requirement is there so it's easier to visually see where the binder starts and ends, you can scan a column to find binders that way.